compiler/src/dotty/tools/dotc/typer/RefChecks.scala

```diff
@@ -936,6 +938,74 @@ object RefChecks {
   }
 
   val NoLevelInfo: RefChecks.OptLevelInfo = new OptLevelInfo()
+
+  /** Verify that references in the user-defined `@implicitNotFound` message are valid.
+   *  (i.e. they refer to a type variable that really occurs in the signature of the annotated symbol.)
+   */
+  private object checkImplicitNotFoundAnnotation:
+
+    /** Warns if the class or trait has an @implicitNotFound annotation
+     *  with invalid type variable references.
+     */
+    def template(sd: SymDenotation)(using Context): Unit =
+      for
+        annotation <- sd.getAnnotation(defn.ImplicitNotFoundAnnot)
+        l@Literal(c: Constant) <- annotation.argument(0)
+      do forEachTypeVariableReferenceIn(c.stringValue) { case (ref, start) =>
+        if !sd.typeParams.exists(_.denot.name.show == ref) then
+          reportInvalidReferences(l, ref, start, sd)
+      }
+
+    /** Warns if the def has parameters with an `@implicitNotFound` annotation
+     *  with invalid type variable references.
+     */
+    def defDef(sd: SymDenotation)(using Context): Unit =
+      for
+        paramSymss <- sd.paramSymss
+        param <- paramSymss
+      do
+        for
+          annotation <- param.getAnnotation(defn.ImplicitNotFoundAnnot)
```

Comment:
Isn't the nested for equivalent to doing `annotation <- param.getAnnotation(defn.ImplicitNotFoundAnnot)` in the outer for directly?

Reply:
Sadly not, because `getAnnotation` returns an option and `paramSymss` returns a `List[List[Symbol]]`